app/controllers/login_controller.rb

```diff
@@ -202,6 +202,13 @@ def twofa_verify
       session[:u] = tmpu.session_token
       session.delete(:twofa_u)
       redirect_to "/"
+    elsif (tmpu = find_twofa_user) && tmpu.authenticate_totp_recovery(params[:totp_code])
+      flash[:error] = "You used a OTP recovery code to log in. Please update your OTP device to ensure continued access to your account."
```

Comment:
Can we give more-specific instructions? Should using a recovery code disable 2fa so they can re-enable it with a new secret?

Reply:
> Should using a recovery code disable 2fa so they can re-enable it with a new secret?

I think this is the more ideal solution.

Reply:
If using a recovery key disables 2fa, and enabling 2fa replaces all the recovery codes, it doesn't seem like there's a benefit to generating more than one. A user could put store their different codes in different backups, but there's no benefit to each having a unique code vs duplicating one code.

This morning GitHub hassled me to view my 16 recovery codes and confirm I had them backed up. They must not disable 2fa on the use of a recovery code, right?

I think my guess about disabling a recovery code was wrong. I don't know what best practices are here and I'm probably not going to reinvent the wheel correctly. @hmadison do you have any kind of guidance we can use here?

Reply:
The best guidance I can find is https://pages.nist.gov/800-63-4/sp800-63b/events/#recovery

Reply:
A recovery key isn't the same thing as a 2nd factor "backup code". I'm assuming because you're generating 6, that it's in the group of static (non-time based) 2fa tokens. These can be used when you don't have access to the otp secret transiently (what I assume github is doing), or to provide a 2nd factor when the otp secret is believed to be unrecoverable. If it's exclusively the latter, then you don't need 6, you should only generate one, and those details about account recovery are more relevant. The account recovery case is a lot harder to implement correctly.